### ISSUE TITLE
[WP 0.3] Problem with vCard export

### ISSUE THREAD

**Issue by micter59** [ltb-project/white-pages#103]
Sorry to annoy you twice in the same day.
I'm also stuck with this.
When I try to export the information for one person in vCard format,
the vCard doesn't contain anything except :
```
BEGIN:VCARD
VERSION:4.0
END:VCARD
```
Here are the log.
I seems for me that the vcard_map is not found (Undefined variable: vcard_map).
Looking at the code in display.php (line 31 and further), it try to define if it's a user_map or a group_map.
The ldap_user_regex is not define, but the ldap_user_base is.

[Thu Mar 11 15:12:48.840231 2021] [php7:notice] [pid 22] [client 172.20.0.1:37654] PHP Notice:  Undefined variable: logout_link in /usr/share/white-pages/htdocs/index.php on line 77, referer: http://localhost:8001/index.php?page=display&dn=cn%3Dabl%2Cou%3Dlabo%2Cou%3Dusers%2Cdc%3Dlps59&search=
[Thu Mar 11 15:12:48.841523 2021] [php7:warn] [pid 22] [client 172.20.0.1:37654] PHP Warning:  Illegal string offset 'count' in /usr/share/white-pages/htdocs/display.php on line 83, referer: http://localhost:8001/index.php?page=display&dn=cn%3Dabl%2Cou%3Dlabo%2Cou%3Dusers%2Cdc%3Dlps59&search=
[Thu Mar 11 15:12:48.841537 2021] [php7:warn] [pid 22] [client 172.20.0.1:37654] PHP Warning:  Illegal string offset 'count' in /usr/share/white-pages/htdocs/display.php on line 83, referer: http://localhost:8001/index.php?page=display&dn=cn%3Dabl%2Cou%3Dlabo%2Cou%3Dusers%2Cdc%3Dlps59&search=
[Thu Mar 11 15:12:48.841542 2021] [php7:warn] [pid 22] [client 172.20.0.1:37654] PHP Warning:  Illegal string offset 'count' in /usr/share/white-pages/htdocs/display.php on line 83, referer: http://localhost:8001/index.php?page=display&dn=cn%3Dabl%2Cou%3Dlabo%2Cou%3Dusers%2Cdc%3Dlps59&search=
[Thu Mar 11 15:12:48.841546 2021] [php7:warn] [pid 22] [client 172.20.0.1:37654] PHP Warning:  Illegal string offset 'count' in /usr/share/white-pages/htdocs/display.php on line 83, referer: http://localhost:8001/index.php?page=display&dn=cn%3Dabl%2Cou%3Dlabo%2Cou%3Dusers%2Cdc%3Dlps59&search=
[Thu Mar 11 15:12:48.841550 2021] [php7:warn] [pid 22] [client 172.20.0.1:37654] PHP Warning:  Illegal string offset 'count' in /usr/share/white-pages/htdocs/display.php on line 83, referer: http://localhost:8001/index.php?page=display&dn=cn%3Dabl%2Cou%3Dlabo%2Cou%3Dusers%2Cdc%3Dlps59&search=
[Thu Mar 11 15:12:48.841554 2021] [php7:warn] [pid 22] [client 172.20.0.1:37654] PHP Warning:  Illegal string offset 'count' in /usr/share/white-pages/htdocs/display.php on line 83, referer: http://localhost:8001/index.php?page=display&dn=cn%3Dabl%2Cou%3Dlabo%2Cou%3Dusers%2Cdc%3Dlps59&search=
[Thu Mar 11 15:12:48.841558 2021] [php7:warn] [pid 22] [client 172.20.0.1:37654] PHP Warning:  Illegal string offset 'count' in /usr/share/white-pages/htdocs/display.php on line 83, referer: http://localhost:8001/index.php?page=display&dn=cn%3Dabl%2Cou%3Dlabo%2Cou%3Dusers%2Cdc%3Dlps59&search=
[Thu Mar 11 15:12:48.841563 2021] [php7:warn] [pid 22] [client 172.20.0.1:37654] PHP Warning:  Illegal string offset 'count' in /usr/share/white-pages/htdocs/display.php on line 83, referer: http://localhost:8001/index.php?page=display&dn=cn%3Dabl%2Cou%3Dlabo%2Cou%3Dusers%2Cdc%3Dlps59&search=
[Thu Mar 11 15:12:48.841567 2021] [php7:warn] [pid 22] [client 172.20.0.1:37654] PHP Warning:  Illegal string offset 'count' in /usr/share/white-pages/htdocs/display.php on line 83, referer: http://localhost:8001/index.php?page=display&dn=cn%3Dabl%2Cou%3Dlabo%2Cou%3Dusers%2Cdc%3Dlps59&search=
[Thu Mar 11 15:12:48.841581 2021] [php7:warn] [pid 22] [client 172.20.0.1:37654] PHP Warning:  Illegal string offset 'count' in /usr/share/white-pages/htdocs/display.php on line 83, referer: http://localhost:8001/index.php?page=display&dn=cn%3Dabl%2Cou%3Dlabo%2Cou%3Dusers%2Cdc%3Dlps59&search=
[Thu Mar 11 15:12:48.841588 2021] [php7:warn] [pid 22] [client 172.20.0.1:37654] PHP Warning:  Illegal string offset 'count' in /usr/share/white-pages/htdocs/display.php on line 83, referer: http://localhost:8001/index.php?page=display&dn=cn%3Dabl%2Cou%3Dlabo%2Cou%3Dusers%2Cdc%3Dlps59&search=
[Thu Mar 11 15:12:48.841592 2021] [php7:warn] [pid 22] [client 172.20.0.1:37654] PHP Warning:  Illegal string offset 'count' in /usr/share/white-pages/htdocs/display.php on line 83, referer: http://localhost:8001/index.php?page=display&dn=cn%3Dabl%2Cou%3Dlabo%2Cou%3Dusers%2Cdc%3Dlps59&search=
[Thu Mar 11 15:12:48.841597 2021] [php7:notice] [pid 22] [client 172.20.0.1:37654] PHP Notice:  Trying to access array offset on value of type int in /usr/share/white-pages/htdocs/display.php on line 83, referer: http://localhost:8001/index.php?page=display&dn=cn%3Dabl%2Cou%3Dlabo%2Cou%3Dusers%2Cdc%3Dlps59&search=
[Thu Mar 11 15:12:48.841601 2021] [php7:warn] [pid 22] [client 172.20.0.1:37654] PHP Warning:  Illegal string offset 'count' in /usr/share/white-pages/htdocs/display.php on line 83, referer: http://localhost:8001/index.php?page=display&dn=cn%3Dabl%2Cou%3Dlabo%2Cou%3Dusers%2Cdc%3Dlps59&search=
[Thu Mar 11 15:12:48.841666 2021] [php7:notice] [pid 22] [client 172.20.0.1:37654] PHP Notice:  Undefined variable: vcard_map in /usr/share/white-pages/htdocs/display.php on line 96, referer: http://localhost:8001/index.php?page=display&dn=cn%3Dabl%2Cou%3Dlabo%2Cou%3Dusers%2Cdc%3Dlps59&search=
[Thu Mar 11 15:12:48.841682 2021] [php7:warn] [pid 22] [client 172.20.0.1:37654] PHP Warning:  Invalid argument supplied for foreach() in /usr/share/white-pages/lib/vcard.inc.php on line 11, referer: http://localhost:8001/index.php?page=display&dn=cn%3Dabl%2Cou%3Dlabo%2Cou%3Dusers%2Cdc%3Dlps59&search=
172.20.0.1 - - [11/Mar/2021:15:12:48 +0100] "GET /index.php?page=display&dn=cn%3Dabl%2Cou%3Dlabo%2Cou%3Dusers%2Cdc%3Dlps59&search=&vcard=1 HTTP/1.1" 200 393 "http://localhost:8001/index.php?page=display&dn=cn%3Dabl%2Cou%3Dlabo%2Cou%3Dusers%2Cdc%3Dlps59&search=" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:86.0) Gecko/20100101 Firefox/86.0" 34 2223 0 22 localhost:80

Don't hesitate to ask for any other information you need.

**Comment by faust64** [comment]
The `vcard_map` being undefined is weird ( https://github.com/ltb-project/white-pages/blob/v0.3/htdocs/display.php#L96 ), as it should be in the default `config.inc.php` ( https://github.com/ltb-project/white-pages/blob/v0.3/conf/config.inc.php#L141 ).
Can you make sure it is defined in your copy? Or share your configuration?

Editing white-pages configuration, it's usually better to keep that file as is, and add your modifications to some `config.inc.local.php`, in the same folder.

The illegal string offset 'count' could be avoided though, I'll look into this.

**Comment by micter59** [comment]
The config.inc.php file is pre-populated with ENV VAR in docker
Variables user_base, user_filter, group_base and group_filter are set there, but redefined in the config.local.inc.php file.

In the config.local.inc.php file, there are 2 vcard_map defined : vcard_user_map and vcard_group_map. I didn't change anything to that in the 2 config files.

**Comment by micter59** [comment]
I've put an error_log() with $type value, at the line 48, where it should be set.
Strangely, it returns "user". So it can find the user at the right place in the LDAP.
Maybe I'm not at the right place in the code to find where the problem is...

By the way, thanks for your help.

**Comment by faust64** [comment]
There shouldn't be `vcard_user_map` nor `vcard_group_map` in your config, if you're running 0.3 code.
Groups vcard would be released with 0.4.
Sounds like you're generating your config.inc.php from the master branch, while running 0.3 code.

You shouldn't change the `config.inc.php` file.
Make sure your container uses its environment variables generating some `config.inc.local.php` instead. That way you're sure to have the right defaults for whichever version you've installed, be that with yum install, git clone, ....

**Comment by micter59** [comment]
OK. I'm using a modified version of this : https://github.com/NoxInmortus/docker-ldap-tool-box
I made a few modification to use a specific version of white-pages, service-desk and self-service-password.
But I forgot to modify the files/white-pages.sh which generate the config.inc.php file !!

I'll look at this, thanks.

**Comment by faust64** [comment]
Right.

https://github.com/NoxInmortus/docker-ldap-tool-box/blob/master/Dockerfile#L26
https://github.com/NoxInmortus/docker-ldap-tool-box/blob/master/Dockerfile#L30
https://github.com/NoxInmortus/docker-ldap-tool-box/blob/master/files/white-pages.sh

So, it clones the master branch by default, then wipes the default configuration.
You probably patched it to clone v0.3?

The Dockerfile should leave the default config alone.
Then create a config.inc.local.php, only re-defining the variables it needs to be set ( https://github.com/NoxInmortus/docker-ldap-tool-box/blob/master/files/white-pages.sh#L32-L41 ).
That would be way simpler to navigate versions. Otherwise, you may have to rewrite that script each time a variable is added, removed, or renamed.

**Comment by micter59** [comment]
Yes, you're right. I patched this to clone a chosen version (the version itself is en ENV VAR).
In my case it's effectively the 0.3.
But I forgot the generation of the config.inc.php file.

It's a bit difficult, because with the image used, the config.inc.local.php file is mounted.
So there's a mix between some parameters set through ENV VAR, and others with the mounted config.local.inc.php.

Anyway, it's not your task. I'll look at this.
For the moment, I've just made modifications to the white-pages.sh script.
But as you said, it will generate other problems in the future. So I won't escape from this without doing the job!

Thanks again for your help.

Michaël
